src/engine/strat_engine/metadata.rs

```diff
@@ -4,262 +4,416 @@
+    pub fn older(&self) -> usize {
+        match self.mdas[0].last_updated.cmp(&self.mdas[1].last_updated) {
+            Ordering::Less => 1,
+            Ordering::Greater => 0,
+            Ordering::Equal => 1,
         }
+    }
 
-        let crc = crc32::checksum_ieee(&buf[offset + 4..offset + SECTOR_SIZE as usize]);
-        if crc != LittleEndian::read_u32(&buf[offset..offset + 4]) {
-            return Err("header CRC invalid".into());
+    pub fn newer(&self) -> usize {
+        match self.older() {
+            0 => 1,
+            1 => 0,
+            _ => panic!("invalid val from older()"),
         }
+    }
+}
 
-        let pool_uuid = try!(Uuid::parse_str(from_utf8(&buf[offset + 32..offset + 64]).unwrap())
-            .map_err(|_| "invalid pool uuid"));
-        let dev_uuid = try!(Uuid::parse_str(from_utf8(&buf[offset + 64..offset + 96]).unwrap())
-            .map_err(|_| "invalid dev uuid"));
+#[derive(Debug, Clone, Copy, Eq, PartialEq)]
+pub struct MDAHeader {
+    pub last_updated: Option<Timespec>,
+    pub used: u64,
+    pub region_size: u64,
+    pub json_crc: u32,
+}
 
-        let mda_size = Sectors(LittleEndian::read_u32(&buf[offset + 160..offset + 164]) as u64);
+impl MDAHeader {
+    pub fn new(region_size: u64) -> MDAHeader {
+        MDAHeader {
+            last_updated: None,
+            used: 0,
+            region_size: region_size,
+            json_crc: 0,
+        }
+    }
+
+    pub fn from_buf(buf: &[u8; MDA_REGION_HDR_SIZE], region_size: u64) -> EngineResult<MDAHeader> {
+        if LittleEndian::read_u32(&buf[..4]) != crc32::checksum_ieee(&buf[4..]) {
+            return Err(EngineError::Engine(ErrorEnum::Invalid, "MDA region header CRC".into()));
+        }
 
-        match validate_mda_size(mda_size) {
-            None => {}
-            Some(err) => {
-                return Err(err.into());
+        let time = {
+            match LittleEndian::read_i64(&buf[16..24]) {
+                0 => None,
+                secs => Some(Timespec::new(secs, LittleEndian::read_i32(&buf[24..28]))),
             }
         };
 
-        Ok(SigBlock {
-            mda: MDAGroup::read(buf, offset + 96, mda_size),
-            pool_uuid: pool_uuid,
-            dev_uuid: dev_uuid,
-            total_size: total_size,
-            mda_sectors: mda_size,
-            reserved_sectors:
-                Sectors(LittleEndian::read_u32(&buf[offset + 164..offset + 168]) as u64),
+        Ok(MDAHeader {
+            used: LittleEndian::read_u64(&buf[8..16]),
+            last_updated: time,
+            region_size: region_size,
+            json_crc: LittleEndian::read_u32(&buf[4..8]),
         })
     }
 
-    /// Read from Stratis magic number area in buf at offset.
-    pub fn read_strat_magic(buf: &[u8], offset: usize) -> &[u8] {
-        &buf[offset + 4..offset + 20]
-    }
+    pub fn to_buf(data: &[u8], timestamp: &Timespec) -> [u8; MDA_REGION_HDR_SIZE] {
+        let mut buf = [0u8; MDA_REGION_HDR_SIZE];
```

Comment:
It feels like you ought to be setting the crc for the header itself in this method.